content/blog/2018-06-07-when-to-use-derived-state.md

```diff
@@ -0,0 +1,384 @@
+
+### Anti-pattern: Props always override state
+
+A common misconception about both `getDerivedStateFromProps` and `componentWillReceiveProps` is that they are only called when props "change". However, any time a parent component re-renders, the child component still receives the new props. Because of this, it is (and has always been) unsafe to _unconditionally_ override state values using either of these lifecycles.
+
+Before React 16.4, it was difficult to reproduce such bugs consistently because they depend on the timing of parent re-rendering. To expose them consistently, we made `getDerivedStateFromProps` in React 16.4 fire for renders caused by `setState` or `forceUpdate` too. Let’s consider an example to demonstrate the problem.
+
+Here is a `EmailInput` component that accepts an email prop:
+```js
+class EmailInput extends Component {
+  state = {
+    email: this.props.email
+  };
+
+  static getDerivedStateFromProps(props, state) {
+    // This is bad! Do not do this!
+    if (props.email) {
+      return { email: props.email };
+    }
+    return null;
+  }
+
+  render() {
+    return <input onChange={this.handleChange} value={this.state.email} />;
+  }
+
+  handleChange = event => {
+    this.setState({ email: event.target.value });
+  };
+}
+```
+
+At first, this component might look okay. State is initialized to the value specified by props and updated when we type into the `<input>`. But once our component re-renders– either because it called `setState` or because its parent re-rendered– anything we've typed into the `<input>` will be lost! (See [this demo for an example](codesandbox://when-to-use-derived-state/derived-state-anti-pattern).)
+
+At this point, you might be wondering if this component would have worked with version 16.3. Unfortunately, the answer is "no". Before moving on, let's take a look at why this is.
+
+For the simple example above, we could "fix" the problem of unexpected re-renders using `shouldComponentUpdate`. However in practice, components usually accept multiple props, and our component would re-render if any one of them changed. Even if none of them "changed", props that are functions or objects are often created inline and so will always bypass `shouldComponentUpdate`. For example, what if our component accepted a function to validate the current email address?
+```js
+<EmailInput
+  email={this.props.user.email}
+  validate={this.validateEmail.bind(this)}
+/>
+```
+
+The above example binds the validation callback inline and so it will pass a new function prop every time it renders– effectively bypassing `shouldComponentUpdate` entirely. [Here is a demo](codesandbox://when-to-use-derived-state/derived-state-anti-pattern) that uses a timer to illustrate this problem. Because it might break at any time your component needs to accept new props, this design pattern is inherently fragile.
```

Comment:
I understand why you did the `prevProps` comparison in that demo link but I think it would be easier to understand if instead you used the same exact code as in the inline example above, but pinned the React version at 16.3.

*Then* it would be a convincing demo of it not working in 16.3 (as advertised above).

Reply:
Yeah. Unfortunately I can't do that with the code sandbox repl plug-in, and I've used all my free sandboxes up 😅 maybe we should temporarily use code pen for this?

Reply:
Yea let's just use CodePen for this post.

Reply:
I've made a new sandbox pinned to 16.3 https://codesandbox.io/s/m3w9zn1z8x